ash/src/vk/prelude.rs

```diff
@@ -59,3 +59,13 @@ impl From<vk::Extent2D> for vk::Rect2D {
 pub unsafe trait TaggedStructure {
     const STRUCTURE_TYPE: vk::StructureType;
 }
+
+#[cfg(feature = "debug")]
+pub(crate) fn wrap_cstr_slice_until_nul(
+    bytes: &[::std::os::raw::c_char],
+) -> Result<&::std::ffi::CStr, core::ffi::FromBytesUntilNulError> {
+    ::std::ffi::CStr::from_bytes_until_nul(
+        // SAFETY: The cast from c_char to u8 is ok because a c_char is always one byte.
+        unsafe { ::std::slice::from_raw_parts(bytes.as_ptr().cast(), bytes.len()) },
+    )
+}
```

Comment:
Dropped the `unwrap()` here so that `Ok("...")` shows up in `Debug`, but at least we won't panic if the struct isn't formattable because the string is invalid.

Reply:
The only kind of error `from_bytes_until_nul()` can return is `FromBytesUntilNulError`, which means it would have gone out of bounds to find a trailing nul (i.e. the given slice doesn't contain a nul character).

Replacement characters are only emitted for `to_string_lossy()` (and `to_str()` returns an error). `Debug`-printing a `CStr`, which is what happens here in `ash`, simply prints the byte-string via `self.to_bytes().escape_ascii()`, which via https://doc.rust-lang.org/std/ascii/fn.escape_default.html prints the ASCII letter or an escape character (e.g. `\x3c`, `\n`, etc).

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=afb5100e527fb9710de21d44cdce3d07

---

So we already get the desired behaviour, just wondering what would be the best way to represent such an invalid string. Since quotation marks are already included I'd fancy if we could either see:

```
device_name: "My GPU",
device_name: FromBytesUntilNulError,
```

Without being wrapped in (multiline) `Err`/`Ok`.

Reply:
Ah, right, `CStr`s aren't UTF-8 to begin with. The proposed formatting seems fine to me.